File: core/src/renderer/camera.cpp
#include "renderer/camera.h"
#include "renderer/draw_command.h"
#include <numbers>
#include <glm/ext/matrix_clip_space.hpp>
#include <glm/ext/matrix_transform.hpp>

const float PI = std::numbers::pi_v<float>;

namespace core
{
glm::mat4 Camera::GetView() const
{
    return glm::lookAt(position, position + direction, up);
}

glm::mat4 Camera::GetProjection() const
{
    switch(projectionType)
    {
    case ProjectionType::ORTHOGRAPHIC: return glm::ortho(-aspect*orthographicHalfHeight, aspect*orthographicHalfHeight, -orthographicHalfHeight, orthographicHalfHeight, near, far);
    case ProjectionType::PERSPECTIVE: return glm::perspective(fovY, aspect, near, far);
    default: break;
    }
    return glm::mat4(1.0f);
}


static CameraSystem* cameraSystem = nullptr;
CameraSystem* GetCameraSystem()
{
    return cameraSystem;
}

void CameraSystem::Begin()
{
    cameraSystem = this;
}

void CameraSystem::Update(float dt)
{
    glm::vec3 direction{};
    if(keys_[0])
    {
        direction += glm::vec3(1,0,0);
    }
    if(keys_[1])
    {
        direction -= glm::vec3(1,0,0);
    }
    if(keys_[2])
    {
        direction += glm::vec3(0,1,0);
    }
    if(keys_[3])
    {
        direction -= glm::vec3(0,1,0);
    }
    const auto horizontal = glm::normalize(glm::cross(camera.up, camera.direction));
    const auto vertical = glm::cross(horizontal, camera.direction);
    const auto windowSize = GetWindowSize();
    const auto rotateX = glm::rotate(glm::mat4(1.0), mouseMotion_.x*PI/static_cast<float>(windowSize.x), vertical);
    const auto rotateY = glm::rotate(glm::mat4(1.0), mouseMotion_.y*PI/static_cast<float>(windowSize.y), horizontal);
    mouseMotion_ = {};
    camera.direction = glm::vec3(rotateX*rotateY*glm::vec4(camera.direction, 1.0));
    camera.position += (horizontal*direction.x+camera.direction*direction.y)*dt;

    //Updating the scene camera
    auto* scene = GetCurrentScene();
    scene->GetCamera() = camera;
}

void CameraSystem::End()
{
    camera = {};
}


void CameraSystem::Draw(DrawCommand* drawCommand)
{
    
}

void CameraSystem::OnKeyDown(SDL_Keycode keycode)
{
    switch(keycode)
    {
    case SDLK_a:
        keys_[0] = true;
        break;
    case SDLK_d:
        keys_[1] = true;
        break;
    case SDLK_w:
        keys_[2] = true;
        break;
    case SDLK_s:
        keys_[3] = true;
        break;
    default:
        break;
    }
}

void CameraSystem::OnKeyUp(SDL_Keycode keycode)
{
    switch(keycode)
    {
    case SDLK_a:
        keys_[0] = false;
        break;
    case SDLK_d:
        keys_[1] = false;
        break;
    case SDLK_w:
        keys_[2] = false;
        break;
    case SDLK_s:
        keys_[3] = false;
        break;
    default:
        break;
    }
}

void CameraSystem::OnMouseMotion(glm::vec2 mouseMotion)
{
    mouseMotion_ = mouseMotion;
}
} // namespace core

File: main/vk_samples/data/scripts/camera.py
import core
import math


class Camera(core.System):
    def begin(self):
        scene = core.get_scene()
        self.camera = scene.get_camera()
        self.camera.position = core.Vec3(0,0,3)
        self.camera.direction = core.Vec3(0,0,-1)
        self.speed = 1.0
        self.keys = [False, False, False, False]
        self.mouse_motion = core.Vec2()

    def update(self, dt: float):
        scene = core.get_scene()
        self.camera = scene.get_camera()
        direction = core.Vec3()
        if self.keys[0]:
            direction += core.Vec3(1, 0, 0)
        if self.keys[1]:
            direction -= core.Vec3(1, 0, 0)
        if self.keys[2]:
            direction += core.Vec3(0, 1, 0)
        if self.keys[3]:
            direction -= core.Vec3(0, 1, 0)
        horizontal = core.Vec3.cross(self.camera.up, self.camera.direction).normalize()
        vertical = core.Vec3.cross(horizontal, self.camera.direction)
        window_size = core.get_window_size()
        self.mouse_motion.x /= window_size.x
        self.mouse_motion.y /= window_size.y
        rotate_x = core.Mat4.rotate(core.Mat4(1.0), self.mouse_motion.x*math.pi, vertical)
        rotate_y = core.Mat4.rotate(core.Mat4(1.0), self.mouse_motion.y*math.pi, horizontal)
        self.mouse_motion = core.Vec2()
        self.camera.direction = core.Vec3(rotate_x*rotate_y*core.Vec4(self.camera.direction, 1.0))

        self.camera.position += (horizontal*direction.x+self.camera.direction*direction.y)*self.speed*dt
        
    def end(self):
        pass

    def on_key_down(self, keycode: core.Key):
        if keycode == int(core.Key.A):
            self.keys[0] = True
        elif keycode == int(core.Key.D):
            self.keys[1] = True
        elif keycode == int(core.Key.W):
            self.keys[2] = True
        elif keycode == int(core.Key.S):
            self.keys[3] = True

    def on_key_up(self, keycode: core.Key):
        if keycode == int(core.Key.A):
            self.keys[0] = False
        elif keycode == int(core.Key.D):
            self.keys[1] = False
        elif keycode == int(core.Key.W):
            self.keys[2] = False
        elif keycode == int(core.Key.S):
            self.keys[3] = False

    def on_mouse_motion(self, motion: core.Vec2):
        self.mouse_motion = motion
